r/aws_instance: add T2 unlimited to aws instance resources

kl4w opened the pull request:
I think this should cover: https://github.com/terraform-providers/terraform-provider-aws/issues/2491

```bash
make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstance_t2Unlimited'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstance_t2Unlimited -timeout 120m
=== RUN   TestAccAWSInstance_t2Unlimited_invalidInstanceType
--- PASS: TestAccAWSInstance_t2Unlimited_invalidInstanceType (24.21s)
=== RUN   TestAccAWSInstance_t2Unlimited_unlimitedCredits
--- PASS: TestAccAWSInstance_t2Unlimited_unlimitedCredits (127.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	151.986s

make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstanceDataSource_t2Unlimited'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstanceDataSource_t2Unlimited -timeout 120m
=== RUN   TestAccAWSInstanceDataSource_t2Unlimited
--- PASS: TestAccAWSInstanceDataSource_t2Unlimited (147.84s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	147.866s
```

kl4w commented:
Yes absolutely, I've been going back and forth about how to implement this portion and I believe your suggestion makes complete sense.

Also considering that the launch template also takes the same structure in its data.

I'll implement the change and get back to you

kl4w commented:
@radeksimko Hopefully this covers your review.

```bash
make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstance_creditSpecification'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstance_creditSpecification -timeout 120m
=== RUN   TestAccAWSInstance_creditSpecification_standardCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits (220.17s)
=== RUN   TestAccAWSInstance_creditSpecification_unlimitedCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits (231.52s)
=== RUN   TestAccAWSInstance_creditSpecification_updateCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_updateCpuCredits (129.54s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	581.273s


make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstanceDataSource_creditSpecification'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstanceDataSource_creditSpecification -timeout 120m
=== RUN   TestAccAWSInstanceDataSource_creditSpecification
--- PASS: TestAccAWSInstanceDataSource_creditSpecification (124.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	124.673s
```

kylelaverty commented:
@radeksimko Was there anything else missing from this PR? I was looking at preping the change for launch configurations for once that update is done but without this implementation confirmed I didn't want to get too far into anything.

kl4w commented on aws/resource_aws_instance.go:
```diff
@@ -1024,6 +1048,26 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
 
+	if d.HasChange("credit_specification") {
+		creditSpecifications := d.Get("credit_specification").([]interface{})
+		if len(creditSpecifications) == 1 {
```
@radeksimko I have a question about HasChange. It looks like it doesn't detect removal of the attributes. Has this always been the case?

It looks like in order to disable the `credit_specification` once it's enabled means setting the value back to `standard`.

What would your recommendation on how to handle this be?

kl4w commented:
@radeksimko aside from your comment on the HasChange, i've taken account of the others, rebased and squashed the commits.

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstance_creditSpecification'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstance_creditSpecification -timeout 120m
=== RUN   TestAccAWSInstance_creditSpecification_standardCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits (166.60s)
=== RUN   TestAccAWSInstance_creditSpecification_unlimitedCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits (171.80s)
=== RUN   TestAccAWSInstance_creditSpecification_updateCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_updateCpuCredits (218.47s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	556.902s
```
```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstanceDataSource_creditSpecification'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstanceDataSource_creditSpecification -timeout 120m
=== RUN   TestAccAWSInstanceDataSource_creditSpecification
--- PASS: TestAccAWSInstanceDataSource_creditSpecification (176.01s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	176.046s
```

dmitrypol commented:
sorry to be a pest but any timeline on when this will be released?

fender4645 commented:
Is there any reason why this hasn't been pulled yet? We're hitting a very obscure yet very real limitation in AWS where you can only spin up ~100 t2 instances with full credits in a 24 hour period and t2.unlimited will solve this for us.

curator commented:
My company's in serious need of Terraform-based support for this, can we please get this pulled?

andycognitoiq commented:
We really want to make use of this too. Please can it be merged and released ASAP.

Thank you!

mrubin commented:
+1 to getting this released.

omeid commented:
@kl4w Could you please rebase this?

appilon commented:
I've actually taken the liberty of rebasing the branch, I'm just running the acceptance tests now. We will see if I can push directly to the branch (allow edits from maintainers).

paultyng commented on aws/resource_aws_instance.go:
```diff
@@ -433,6 +433,22 @@ func resourceAwsInstance() *schema.Resource {
+						"cpu_credits": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
```
Should this have validation? I'm not familiar with the API, but I only seen `unlimited` and `standard` in the results? Is it really free form?

omeid replied:
I believe the API complains if you use something other than `unlimited` and `standard`, leaving this opaque in terraform gives _out of the box_ support for new `cpu_credit` types in the future.

appilon commented:
@kl4w I took the liberty rebasing your branch (I only solved the merge conflicts I didn't change anything), this means you will need to do a reset locally. I'm new to the TF team so I might have jumped the gun on rebasing your work to try and get it across the finish line, my sincerest apologies if that's the case. Feel free to ping me if there is anything I can do on this PR going forward. Otherwise amazing work 🎉

kl4w commented:
@appilon that’s perfect, thanks. I’m actually on break and wouldn’t have been able to get to this for a few more days.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

jmehnle commented:
So what needs doing next? Any way I can help to get this merged?

rbnrye commented:
+1, this is a critical feature for me too. Would be great if someone could advice on why this has been  pushed from milestone 1.12 to 1.13 to 1.14, and/or if there is anything that needs doing?

leigh507 commented:
+1 - Need to move away from m3.medium - and the next closest (C5.Large) is pretty expensive. Using t2.unlimited with a t2.medium would be a great way to solve this.

dlyub commented:
Waiting for this feature to become available as well, any way to get it in sooner?

GoodMirek commented:
This feature is pretty important for me too. I do not understand why this feature is pushed back and why related comments are marked as off-topic.

rbnrye commented:
Can't tell who marked the comments as off-topic, but maybe @bflad can elaborate on why it keeps getting pushed out to the next milestone when seemingly, from this pull requests, all is ready to go?

bflad commented:
@kl4w we just ran into a similar situation with `aws_codebuild_project` and its `cache.0.type` attribute -- basically we opted for:

Removing `Computed: true` from parent and child attribute, adding a `Default: X` to the child attribute, and adding a `DiffSuppressFunc` on the parent attribute to prevent the perpetual difference when not defined in the configuration:

https://github.com/terraform-providers/terraform-provider-aws/blob/7bcafda1f20b56838e514e68148af11d2e3d8190/aws/resource_aws_codebuild_project.go#L74-L79

Then using this behavior for handle the update when it was removed:

https://github.com/terraform-providers/terraform-provider-aws/blob/7bcafda1f20b56838e514e68148af11d2e3d8190/aws/resource_aws_codebuild_project.go#L577-L585

Maybe worth giving this a shot here too?

bflad commented on aws/data_source_aws_instance.go:
```diff
@@ -389,6 +401,13 @@ func instanceDescriptionAttributes(d *schema.ResourceData, instance *ec2.Instanc
+		if err != nil {
+			return err
+		}
+		d.Set("credit_specification", creditSpecifications)
```
We should return errors when attempting to set non-scalar attributes in the state (otherwise it will silently use the value of the configuration or in the case of the data source, not set it at all):

```go
if err := d.Set("credit_specification", creditSpecifications); err != nil {
  return fmt.Errorf("error setting credit_specification: %s", err)
}
```

bflad commented on aws/resource_aws_instance.go:
```diff
@@ -783,6 +800,13 @@ func resourceAwsInstanceRead(d *schema.ResourceData, meta interface{}) error {
+		if err != nil {
+			return err
+		}
+		d.Set("credit_specification", creditSpecifications)
```
Same here -- we should return errors when attempting to set non-scalar attributes in the state (otherwise it will silently use the value of the configuration or in the case of the data source, not set it at all):

```go
if err := d.Set("credit_specification", creditSpecifications); err != nil {
  return fmt.Errorf("error setting credit_specification: %s", err)
}
```

bflad commented on aws/resource_aws_instance.go:
```diff
@@ -1644,6 +1689,18 @@ func buildAwsInstanceOpts(
 
+	if v, ok := d.GetOk("credit_specification"); ok {
+		cs := v.([]interface{})
+		for _, csValue := range cs {
```
Since there will only be one element if its defined, we can skip the loop and directly access `cs[0]`

bflad commented on aws/resource_aws_instance_test.go:
```diff
@@ -1397,6 +1397,78 @@ func TestAccAWSInstance_getPasswordData_trueToFalse(t *testing.T) {
+					resource.TestCheckResourceAttr(resName, "credit_specification.#", "1"),
+					resource.TestCheckResourceAttr(resName, "credit_specification.0.cpu_credits", "unlimited"),
+				),
+			},
```
Please add a `TestStep` that removes the `credit_specification` configuration and expects to find `credit_specification.0.cpu_credits` set to `standard`

kl4w commented:
@bflad I've rebased (from some time yesterday iirc) and made changes based on your review

```
TF_SCHEMA_PANIC_ON_ERROR=1 make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstance_creditSpecification_'
gofmt -w $(find . -name '*.go' |grep -v vendor)
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstance_creditSpecification_ -timeout 120m
=== RUN   TestAccAWSInstance_creditSpecification_unspecifiedDefaultsToStandard
--- PASS: TestAccAWSInstance_creditSpecification_unspecifiedDefaultsToStandard (97.88s)
=== RUN   TestAccAWSInstance_creditSpecification_standardCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits (106.56s)
=== RUN   TestAccAWSInstance_creditSpecification_unlimitedCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits (89.98s)
=== RUN   TestAccAWSInstance_creditSpecification_updateCpuCredits
--- PASS: TestAccAWSInstance_creditSpecification_updateCpuCredits (113.34s)
=== RUN   TestAccAWSInstance_creditSpecification_removalReturnsStandard
--- PASS: TestAccAWSInstance_creditSpecification_removalReturnsStandard (88.56s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	496.363s
```

However I'm getting some failures on tests (even on master) with some aws_instance (both resource and data resource). How would you like me to proceed?

bflad commented:
> However I'm getting some failures on tests (even on master and also unrelated to my changes) with some aws_instance (both resource and data resource). How would you like me to proceed?

Are they related to the `aws_security_group` and `aws_route` resources? They are known to be flakey due to EC2 eventual consistency. Some are already noted in bug reports across this repository. It'd be helpful to know what you're seeing to make a determination though.

I'll kick off a full acceptance test run now and ensure nothing here is breaking backwards compatibility.

kl4w commented:
Yeah, both of those resources. Also, this is on latest master (all the data resource tests fail)

```
TF_SCHEMA_PANIC_ON_ERROR=1 make testacc TEST=./aws TESTARGS='-run=TestAccAWSInstanceDataSource_basic'gofmt -w $(find . -name '*.go' |grep -v vendor)
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSInstanceDataSource_basic -timeout 120m
=== RUN   TestAccAWSInstanceDataSource_basic
panic: Invalid address to set: []string{"disable_api_termination"}
```

bflad commented:
> `panic: Invalid address to set: []string{"disable_api_termination"}`

We'll address this separately 😉

radeksimko commented:
Just to reiterate on a couple of questions in this thread, specifically from @GoodMirek and @rbnrye :

I did mark a couple of comments as off-topic and didn't provide further explanation - apologies for that. I'll be sure to do this in future where these comments are non-trivial.

Generally speaking using reactions is always a preferred way to express interest (or other emotions) as we can use these when sorting PRs and issues easily.

Comments along the lines of "+1" or "when will this ship" whilst seeming helpful send everybody involved a notification which causes issues across a larger repository (such that we ask that folks use the reactions to indicate you're interested in a feature). We're taking advantage of the new "off-topic" feature in PR's and Issues to try and make it clearer what the remaining state of the PR is (e.g. any comments from PR review, discussions about how it should be implemented etc.).

TL;DR: We do appreciate people expressing interest in PRs and helping us prioritise, but comments are not the best way to do so, for the reasons mentioned.

bflad commented:
This has been released in [version 1.16.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1160-april-25-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

jleonar commented:
I just wanted to let you know that this change breaks deployments into the Govcloud region of us-gov-west-1.

Haven't tested elsewhere.

This is the error that I receive.  

`Response ec2/DescribeInstanceCreditSpecifications Details:
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: ---[ RESPONSE ]--------------------------------------
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: HTTP/1.1 400 Bad Request
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: Connection: close
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: Transfer-Encoding: chunked
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: Date: Thu, 26 Apr 2018 18:17:02 GMT
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: Server: AmazonEC2
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: 
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: 
2018-04-26T14:17:02.994-0400 [DEBUG] plugin.terraform-provider-aws_v1.16.0_x4: -----------------------------------------------------`

bflad commented:
The fix for GovCloud and China (or really anything throwing `UnsupportedOperation` with the new call) was made in #4362, merged into master, and will release with v1.17.0 of the AWS provider middle of next week.
